File: gateway-service/internal/models/monitoring/jaeger.go
package monitoring

import (
	"context"
	"crypto/tls"
	"fmt"
	"log/slog"
	"time"

	monpb "gateway-service/internal/gen/proto/go/vartrack/v1/models/monitoring"
	"gateway-service/internal/protoutil"

	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/encoding/gzip"
)

// JaegerBackend ships spans to a Jaeger collector via OTLP/gRPC or OTLP/HTTP.
//
// Implements:
//   - Backend            (Name / Ping / Shutdown)
//   - tracerProviderBackend  (TracerProvider)
//   - propagatorBackend      (Propagator — installs W3C TraceContext + Baggage)
//
// When monitoring.Init() receives this backend it installs the TracerProvider
// globally, which makes otelhttp and otelgrpc auto-instrumentation work with
// zero additional configuration at call sites.
type JaegerBackend struct {
	cfg      *monpb.JaegerConfig
	provider *sdktrace.TracerProvider
}

// NewJaegerBackend creates and starts the Jaeger exporter pipeline.
// Returns a disabled no-op backend (not an error) when enabled: false.
func NewJaegerBackend(ctx context.Context, cfg *monpb.JaegerConfig) (*JaegerBackend, error) {
	b := &JaegerBackend{cfg: cfg}

	if !cfg.GetEnabled() {
		slog.Info("jaeger: disabled")
		return b, nil
	}

	tlsCfg, err := buildBackendTLS(
		cfg.GetTlsCaCert(), cfg.GetTlsClientCert(), cfg.GetTlsClientKey(),
		cfg.GetInsecureSkipVerify(),
	)
	if err != nil {
		return nil, fmt.Errorf("jaeger: TLS: %w", err)
	}

	exp, err := b.buildExporter(ctx, tlsCfg)
	if err != nil {
		return nil, fmt.Errorf("jaeger: exporter: %w", err)
	}

	res, err := buildOTelResource(ctx,
		cfg.GetServiceName(), cfg.GetServiceVersion(), cfg.GetEnvironment(),
		cfg.GetResourceAttributes(),
	)
	if err != nil {
		return nil, fmt.Errorf("jaeger: resource: %w", err)
	}

	batchOpts := []sdktrace.BatchSpanProcessorOption{}
	if v := cfg.GetMaxQueueSize(); v > 0 {
		batchOpts = append(batchOpts, sdktrace.WithMaxQueueSize(int(v)))
	}
	if d := protoutil.DurationOrDefault(cfg.GetFlushInterval(), 0); d > 0 {
		batchOpts = append(batchOpts, sdktrace.WithBatchTimeout(d))
	}

	b.provider = sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exp, batchOpts...),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(buildJaegerSampler(cfg)),
	)

	slog.Info("jaeger: started",
		"endpoint", cfg.GetEndpoint(),
		"protocol", cfg.GetProtocol(),
		"service", cfg.GetServiceName(),
		"sampler", cfg.GetSamplerType(),
	)
	return b, nil
}

func (b *JaegerBackend) Name() string { return "jaeger" }

// TracerProvider satisfies tracerProviderBackend.
// monitoring.Init() installs this as the global OTel TracerProvider.
func (b *JaegerBackend) TracerProvider() trace.TracerProvider {
	if b.provider == nil {
		return nil
	}
	return b.provider
}

// Propagator installs W3C TraceContext + Baggage so otelhttp reads the incoming
// traceparent header and otelgrpc injects it into outbound calls.
func (b *JaegerBackend) Propagator() propagation.TextMapPropagator {
	return propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	)
}

// Ping force-flushes buffered spans.  Used by the background health logger.
func (b *JaegerBackend) Ping(ctx context.Context) error {
	if b.provider == nil {
		return nil
	}
	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()
	if err := b.provider.ForceFlush(ctx); err != nil {
		return fmt.Errorf("jaeger ping: %w", err)
	}
	return nil
}

// Shutdown flushes remaining spans and closes the exporter.
func (b *JaegerBackend) Shutdown(ctx context.Context) error {
	if b.provider == nil {
		return nil
	}
	if err := b.provider.Shutdown(ctx); err != nil {
		return fmt.Errorf("jaeger shutdown: %w", err)
	}
	slog.Info("jaeger: shut down")
	return nil
}

// ── exporter factory ──────────────────────────────────────────────────────────

func (b *JaegerBackend) buildExporter(ctx context.Context, tlsCfg *tls.Config) (sdktrace.SpanExporter, error) {
	switch b.cfg.GetProtocol() {
	case "grpc", "":
		opts := []otlptracegrpc.Option{
			otlptracegrpc.WithEndpoint(b.cfg.GetEndpoint()),
			otlptracegrpc.WithCompressor(gzip.Name), // compress spans; saves bandwidth at scale
		}
		if b.cfg.GetUseTls() {
			opts = append(opts, otlptracegrpc.WithTLSCredentials(credentials.NewTLS(tlsCfg)))
		} else {
			opts = append(opts, otlptracegrpc.WithDialOption(
				grpc.WithTransportCredentials(insecure.NewCredentials()),
			))
		}
		if t := b.cfg.GetAuthToken(); t != "" {
			opts = append(opts, otlptracegrpc.WithHeaders(map[string]string{
				"Authorization": "Bearer " + t,
			}))
		}
		return otlptracegrpc.New(ctx, opts...)

	case "thrift_http":
		opts := []otlptracehttp.Option{
			otlptracehttp.WithEndpoint(b.cfg.GetEndpoint()),
		}
		if !b.cfg.GetUseTls() {
			opts = append(opts, otlptracehttp.WithInsecure())
		}
		if t := b.cfg.GetAuthToken(); t != "" {
			opts = append(opts, otlptracehttp.WithHeaders(map[string]string{
				"Authorization": "Bearer " + t,
			}))
		}
		return otlptracehttp.New(ctx, opts...)

	default:
		return nil, fmt.Errorf("unsupported protocol %q — use grpc or thrift_http", b.cfg.GetProtocol())
	}
}

// ── sampler mapping ───────────────────────────────────────────────────────────

func buildJaegerSampler(cfg *monpb.JaegerConfig) sdktrace.Sampler {
	switch cfg.GetSamplerType() {
	case "const":
		if cfg.GetSamplerParam() == 0 {
			return sdktrace.NeverSample()
		}
		return sdktrace.AlwaysSample()
	case "probabilistic":
		return sdktrace.ParentBased(sdktrace.TraceIDRatioBased(cfg.GetSamplerParam()))
	case "rateLimiting":
		// OTel SDK has no rate-limiting sampler; approximate with ratio (param = req/s ÷ 1000).
		return sdktrace.ParentBased(sdktrace.TraceIDRatioBased(cfg.GetSamplerParam() / 1000))
	default:
		return sdktrace.ParentBased(sdktrace.AlwaysSample())
	}
}

File: orchestrator-service/app/monitoring/jaeger.py
"""
app/monitoring/jaeger.py
─────────────────────────
Jaeger monitoring backend (OTLP/gRPC or OTLP/HTTP).

Ships spans to a Jaeger collector.  Mirrors gateway's
models/monitoring/jaeger.go JaegerBackend.

Requires: opentelemetry-sdk opentelemetry-exporter-otlp-proto-grpc
          (or opentelemetry-exporter-otlp-proto-http for thrift_http)
"""
from __future__ import annotations

import logging

from app.monitoring.base import Backend
from app.monitoring._utils import _build_ssl_context, _build_otel_resource, _build_otel_sampler
from app.monitoring.config import JaegerConfig

logger = logging.getLogger(__name__)


class JaegerBackend(Backend):
    """
    Ships spans to a Jaeger collector via OTLP/gRPC or OTLP/HTTP.
    Mirrors gateway's models/monitoring/jaeger.go JaegerBackend.

    When opentelemetry-sdk is not installed, this backend is a no-op
    (same degradation as gateway's Span falling back to slog entries).
    """

    def __init__(self, cfg: JaegerConfig, version: str, commit: str) -> None:
        self._cfg      = cfg
        self._provider = None

        if not cfg.enabled:
            logger.info("jaeger backend: disabled")
            return

        try:
            self._provider = self._build_provider(cfg, version)
            logger.info(
                "jaeger backend: started endpoint=%s protocol=%s service=%s sampler=%s",
                cfg.endpoint, cfg.protocol, cfg.service_name, cfg.sampler_type,
            )
        except ImportError:
            logger.warning(
                "jaeger backend: opentelemetry-sdk not installed — traces disabled. "
                "pip install opentelemetry-sdk opentelemetry-exporter-otlp-proto-grpc"
            )
        except Exception:
            logger.exception("jaeger backend: startup failed")

    def name(self) -> str:
        return "jaeger"

    def ping(self) -> None:
        if self._provider is None:
            return
        try:
            self._provider.force_flush(timeout_millis=5000)
        except Exception as exc:
            raise RuntimeError(f"jaeger ping: {exc}") from exc

    def shutdown(self) -> None:
        if self._provider is not None:
            try:
                self._provider.shutdown()
            except Exception:
                logger.warning("jaeger backend: shutdown error", exc_info=True)
        logger.info("jaeger backend: shut down")

    @staticmethod
    def _build_provider(cfg: JaegerConfig, version: str):
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor

        resource = _build_otel_resource(
            cfg.service_name, cfg.service_version or version,
            cfg.environment, cfg.resource_attributes,
        )
        sampler = _build_otel_sampler(cfg.sampler_type, cfg.sampler_param)

        tp_kwargs: dict = {}
        if resource:
            tp_kwargs["resource"] = resource
        if sampler:
            tp_kwargs["sampler"] = sampler

        provider = TracerProvider(**tp_kwargs)

        exporter = JaegerBackend._build_exporter(cfg)
        bsp_kwargs: dict = {}
        if cfg.max_queue_size > 0:
            bsp_kwargs["max_queue_size"] = cfg.max_queue_size
        if cfg.flush_interval_seconds > 0:
            bsp_kwargs["schedule_delay_millis"] = int(cfg.flush_interval_seconds * 1000)

        provider.add_span_processor(BatchSpanProcessor(exporter, **bsp_kwargs))

        # Install as global so otelhttp / otelgrpc auto-instrumentation works.
        from opentelemetry import trace
        trace.set_tracer_provider(provider)

        # W3C TraceContext + Baggage propagation (same as gateway's Propagator()).
        from opentelemetry import propagate
        from opentelemetry.propagators.composite import CompositePropagator
        from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
        from opentelemetry.baggage.propagation import W3CBaggagePropagator
        propagate.set_global_textmap(
            CompositePropagator([TraceContextTextMapPropagator(), W3CBaggagePropagator()])
        )

        return provider

    @staticmethod
    def _build_exporter(cfg: JaegerConfig):
        if cfg.protocol in ("grpc", ""):
            from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
            kwargs: dict = {"endpoint": cfg.endpoint}
            if cfg.use_tls:
                import grpc
                _build_ssl_context(cfg.tls_ca_cert, cfg.tls_client_cert, cfg.tls_client_key, cfg.insecure_skip_verify)
                kwargs["credentials"] = grpc.ssl_channel_credentials()
            else:
                kwargs["insecure"] = True
            if cfg.auth_token:
                kwargs["headers"] = {"Authorization": f"Bearer {cfg.auth_token}"}
            return OTLPSpanExporter(**kwargs)

        # thrift_http → OTLP/HTTP
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter as HTTPExporter
        kwargs = {"endpoint": f"{'https' if cfg.use_tls else 'http'}://{cfg.endpoint}/v1/traces"}
        if cfg.auth_token:
            kwargs["headers"] = {"Authorization": f"Bearer {cfg.auth_token}"}
        return HTTPExporter(**kwargs)
